#2992 Prevent XSS for body request

Patrykb0802 opened the pull request:
Added:
- OWASP validator for requests bodies
- Annotation for XssValid.java (used for custom CSS and user comments in alarms section)
- Tests for validator class

Patrykb0802 committed:
#2992 Prevent XSS for body request

Added:
- OWASP validator for requests bodies
- Annotation for XssValid.java (used for custom CSS and user comments in alarms section)
- Tests for validator class

Limraj commented on src/org/scada_lts/web/mvc/api/UserCommentAPI.java:
```diff
@@ -98,6 +99,8 @@ public ResponseEntity<String> deleteUserComment(HttpServletRequest request, @Pat
 
     static class CreateUserComment {
+
+        @XssValid
```
XssProtect


Something is missing in the api method.

Limraj commented on src/org/scada_lts/web/security/OwaspXssValidator.java:
```diff
@@ -0,0 +1,36 @@
+        }
+    }
+
+    void validateXss(String input) throws XssValidatorException {
```
removed

Limraj commented on src/org/scada_lts/web/security/OwaspXssValidator.java:
```diff
@@ -0,0 +1,36 @@
+        } catch (Exception e) {
+            throw new XssValidatorException("XSS validation failed: " + e.getMessage(), e);
+        }
+    }
```
```

@Override
public void validate(String input) throws XssValidatorException {
       String sanitized = policyFactory.sanitize(input);
        if (!sanitized.equals(input)) {
            throw new XssValidatorException("Potential XSS attack detected");
        }
}
```

Limraj commented on src/org/scada_lts/web/security/OwaspXssValidator.java:
```diff
@@ -0,0 +1,36 @@
+
+    private PolicyFactory createPolicyFactory() {
+        return new HtmlPolicyBuilder()
+                .allowElements("b", "i", "u", "strong", "em", "p", "ul", "ol", "li")
```
It can't be like this, there should be no such restrictions.
Does this library actually check text for xss attacks?

Limraj commented on test/org/scada_lts/web/security/XssUtilsValidateHttpBodyTest.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Collection;
+
+@RunWith(Parameterized.class)
+public class XssUtilsValidateHttpBodyTest {
```
Where does this name come from?

Limraj commented on src/org/scada_lts/web/security/XssValidatorException.java:
```diff
@@ -0,0 +1,25 @@
+package org.scada_lts.web.security;
```
As I said, you throw it into the package:
org.scada_lts.web.beans.validation.xss

There are validations using the Java Bean Validation API mechanism.

Limraj reviewed (commented):
What is the source of the library used?

Patrykb0802 commented:
> What is the source of the library used?

https://owasp.org/www-project-java-html-sanitizer/

Patrykb0802 committed:
#2992 Prevent XSS for body request

Added:
- changed PolicyRefactor for OwaspXssValidator.java
- Moved files to different packages (refactor)

Patrykb0802 committed:
#2992 Prevent XSS for body request

Patrykb0802 committed:
#2992 Prevent XSS for body request

Patrykb0802 committed:
#2992 Prevent XSS for body request

Patrykb0802 committed:
#2992 Prevent XSS for body request

Patrykb0802 committed:
#2992 Prevent XSS for body request

Patrykb0802 committed:
#2992 Prevent XSS for body request

Limraj committed:
#2992 Prevent XSS for body request:
- corrected xss body validation - use regex;
- added tests: BodyXssUtilsTest, XssUtilsTestsSuite;
- added cases to tests: XssValidatorExceptionTest, XssValidatorTest;
- added properties in env.properties:
scadalts.security.http.body.protect.enabled
scadalts.security.http.body.access.denied.regex
scadalts.security.http.body.access.granted.regex;
- refactored XssUtils: added methods: validate, validateHttpBody;

Limraj committed:
#2992 Prevent XSS for body request:
- corrected name for class XssValidator;

Limraj committed:
#2992 Prevent XSS for body request:
- corrected systemSettings.jsp;

Limraj committed:
#2992 Prevent XSS for body request:
- corrected granted regex;

Limraj committed:
#2992 Prevent XSS for body request:
- added cases test to BodyXssUtilsTest;
